src/chapter22/exercise01.lean

```diff
@@ -35,76 +41,424 @@ begin
   norm_num,
 end
 
--- if you proved there was an upper bound, then change 37 to the 
--- right answer and prove it's a least upper bound!
--- If you proved there was no upper bound, just comment out
--- this part.
-lemma part_a_lub : is_lub Si 37 :=
+-- 7 is the least upper bound.
+lemma part_a_lub : is_lub Si 7 :=
 begin
-  sorry
+  rw is_lub,
+  rw is_least,
+  split,
+  rw upper_bounds,
+  intro t,
+  intro h,
+  cases h,
+  rw h,
+  norm_num,
+  cases h,
+  rw h,
+  norm_num,
+  cases h,
+  rw h,
+  simp at h,
+  rw h,
+  norm_num,
+  rw lower_bounds,
+  intro n,
+  intro h,
+  rw upper_bounds at h,
+  simp at h,
+  apply h,
+  unfold Si,
+  simp,
 end
 
--- Now the same for lower bounds.
+-- Si is bounded below.
 lemma part_a_lb : ∃ z : ℝ, z ∈ lower_bounds Si :=
 begin
-  sorry
+  use -2,
+  rw lower_bounds,
+  intro t,
+  intro h,
+  cases h,
+  rw h,
+  norm_num,
+  cases h,
+  rw h,
+  norm_num,
+  cases h,
+  rw h,
+  norm_num,
+  simp at h,
+  rw h,
 end
 
--- if you proved there *was* a lower bound,
--- and prove it's a least upper bound!
--- If not, then delete this part.
-lemma part_a_glb : is_glb Si 37 :=
+-- -2 is the greatest lower bound.
+lemma part_a_glb : is_glb Si (-2) :=
 begin
-  sorry
+  rw is_glb,
+  rw is_greatest,
+  split,
+  rw lower_bounds,
+  intro t,
+  intro h,
+  cases h,
+  rw h,
+  norm_num,
+  cases h,
+  rw h,
+  norm_num,
+  cases h,
+  rw h,
+  norm_num,
+  simp at h,
+  rw h,
+  rw upper_bounds,
+  intro n,
+  intro h,
+  rw lower_bounds at h,
+  simp at h,
+  apply h,
+  unfold Si,
+  simp,
 end
 
 def Sii : set ℝ := {x | ∥x - 3∥ < ∥x + 7∥}.
 
--- If you think Sii is bounded above, prove this. If you
--- think it's not, then *disprove* it by putting `¬` in front of it.
-lemma part_b_ub : ∃ z : ℝ, z ∈ upper_bounds Sii :=
+-- Sii is not bounded above.
+lemma part_b_ub : ¬ ∃ z : ℝ, z ∈ upper_bounds Sii :=
 begin
-  sorry
+  rw not_exists,
+  intro n,
+  rw upper_bounds,
+  simp,
+  by_cases h: n ≤ 3,
+  use 4,
+  split,
+  unfold Sii,
+  simp,
+  norm_num,
+  rw abs_of_pos,
+  norm_num,
+  norm_num,
+  linarith,
+  use n+1,
+  split,
```

Comment:
This is a bit annoying because now you have two goals so you should be using `{ }` braces, but the second goal is really easy, it's going to be a lemma in the library. For that second goal (n < n + 1) I used `library_search` to discover that it was called `lt_add_one n`, so instead of `split` you can do
`  refine ⟨_, lt_add_one n⟩,` and you still only have one goal.